design-proposals/host-ip-pool.md

```diff
@@ -0,0 +1,503 @@
+- 2 -> 2001:4860:4860::0010 (excluded)
+...
+- 18 -> 2001:4860:4860::001F (excluded)
+- 19 -> 2001:4860:4860::0020
+...
+- 240 -> 2001:4860:4860::00FF
+
+Reseting `status.allocated-offset` field will trigger a Reconcile on the
+`NodeIPPool` and `allocated-offsets` will be re-calculated.
+In this process the reason will be "ReconcilingPool" and `Available`
+marked as unknown, so users can wait on it.
+
+Following is a table with the "type=Available" condition states
+
+| reason              | status  | message |
+| --------------------| --------| ------ |
+| ReconcilingPool     | Unknown | The pool is being updated |
+| ExhaustedIPPool     | False   | The IP pool has being depleted, manual intervention needed to modify NodeIPPool.|
+| UnsyncronizedIPPool | False   | Failed re constructing the IP pool, it may be not in sync with nodes. |
+| OverlapingIPPool    | False   | The NodeIPPool XYZ is overlapping with this one. |
+| IPCollision         | False   | The IP address XYZ at node XYZ is is colliding with node XYZ |
+
+To assign an IP address from a pool to an interface using NNCP, a special
+capture syntax is introduced understood by kubernetes-nmstate
+`ippool.[name].allocate` that will trigger the ip allocation mechanism.
+It will inject the result on the captured state passed to nmpolicy so it can be
+consumed by it, the `ippool.[name].allocate` tag will return a struct with the
+following fields `ip` and `prefix-length`.
+
+```yaml
+apiVersion: nmstate.io/v1
+kind: NodeNetworkConfigurationPolicy
+metadata:
+  name: static-ip
+capture:
+    address-from-traffic-1: ippool.traffic-1.allocate
+    address-from-traffic-2: ippool.traffic-2.allocate
+desiredState:
+    interfaces:
+    - name: eth1
+      type: ethernet
+      state: up
+      ipv4:
+        address:
+        - ip: "{{ capture.address-from-traffic-1.ip }}"
+          prefix-length: "{{ capture.address-from-traffic-1.prefix-length }}"
+        dhcp: false
+        enabled: true
+    - name: eth2
+      type: ethernet
+      state: up
+      ipv4:
+        address:
+        - ip: "{{ capture.address-from-traffic-2.ip }}"
+          prefix-length: "{{ capture.address-from-traffic-2.prefix-length }}"
+        dhcp: false
+        enabled: true
+```
+
+The allocated IPs will be marked at the `NodeNetworkConfigurationEnactment`
+status and will be used to re-construct the nmstate on reboot and to discover
+if new allocation if needed or to de-allocate them the pool is no longer
+referenced, this is done on distributed fashion by each handler at their nodes.
+
+The `NNCE.status.ippool` information allows reserve those IPs again if the
+`NodeIPPool` is re-created after exhaustion, since fields on the `NodeIPPool.spec` section
+are read-only for simplicity.
+
+If the IPs reserved for a node for a specific pool want to be deallocated the
+user can change `NNCE.status.ippool` fields if the node is Ready the IP allocation
+will try to get a new one for the desiredState.
```

Comment:
This doesn't read well: I don't think IPs want to be deallocated. They don't want anything :P 

```suggestion
The user can change the `NNCE.status.ippool` attribute to deallocate the IPs reserved for a node.

If the node is Ready,  the nmstate will attempt to get a new IP allocation for the desiredState.
```